#!/bin/sh /usr/share/dpatch/dpatch-run
## 02-path-info.dpatch by Evan Jones <ejones@uwaterloo.ca>
##
## DP: Fix illegal PATH_INFO (Closes: #164306).

@DPATCH@

diff -Naur thttpd-2.25b.orig/libhttpd.c thttpd-2.25b/libhttpd.c
--- thttpd-2.25b.orig/libhttpd.c	2007-10-14 10:09:55.000000000 +0000
+++ thttpd-2.25b/libhttpd.c	2007-10-14 10:09:11.000000000 +0000
@@ -1492,8 +1492,6 @@
     restlen = strlen( path );
     httpd_realloc_str( &rest, &maxrest, restlen );
     (void) strcpy( rest, path );
-    if ( rest[restlen - 1] == '/' )
-	rest[--restlen] = '\0';         /* trim trailing slash */
     if ( ! tildemapped )
 	/* Remove any leading slashes. */
 	while ( rest[0] == '/' )
@@ -2322,8 +2320,11 @@
 	{
 	int i;
 	i = strlen( hc->origfilename ) - strlen( hc->pathinfo );
-	if ( i > 0 && strcmp( &hc->origfilename[i], hc->pathinfo ) == 0 )
-	    hc->origfilename[i - 1] = '\0';
+	if ( strcmp( &hc->origfilename[i], hc->pathinfo ) == 0 )
+	    {
+	    if ( i == 0 ) hc->origfilename[0] = '\0';
+	    else hc->origfilename[i - 1] = '\0';
+	    }
 	}
 
     /* If the expanded filename is an absolute path, check that it's still
